knowledge/knowledge_read.go

```diff
@@ -0,0 +1,243 @@
+// by its externalID + type combination.
+func (c *Client) GetDigitalTwinByIdentifier(
+	ctx context.Context,
+	identifier *Identifier,
+	opts ...grpc.CallOption,
+) (*knowledgepb.Node, error) {
+	return c.getNodeByIdentifier(ctx, DigitalTwin, identifier, opts...)
+}
+
+// GetResourceByID is a helper function that queries for a Resource node by its id.
+func (c *Client) GetResourceByID(
+	ctx context.Context,
+	id string,
+	opts ...grpc.CallOption,
+) (*knowledgepb.Node, error) {
+	ctx = insertMetadata(ctx, c.xMetadata)
+	return c.getNodeByID(ctx, id, Resource, opts...)
+}
+
+// GetResourceByIdentifier is a helper function that queries for a Resource node
+// by its externalID + type combination.
+func (c *Client) GetResourceByIdentifier(
+	ctx context.Context,
+	identifier *Identifier,
+	opts ...grpc.CallOption,
+) (*knowledgepb.Node, error) {
+	ctx = insertMetadata(ctx, c.xMetadata)
+	return c.getNodeByIdentifier(ctx, Resource, identifier, opts...)
+}
+
+// ListDigitalTwinsByProperty is a helper function that lists all DigitalTwin nodes
+// that have the specified property.
+func (c *Client) ListDigitalTwinsByProperty(
+	ctx context.Context,
+	property *knowledgepb.Property,
+	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
+	return c.listNodesByProperty(ctx, DigitalTwin, property, opts...)
+}
+
+// ListResourcesByProperty is a helper function that lists all Resource nodes.
+// that have the specified property.
+func (c *Client) ListResourcesByProperty(
+	ctx context.Context,
+	property *knowledgepb.Property,
+	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
+	return c.listNodesByProperty(ctx, Resource, property, opts...)
+}
+
+// ListDigitalTwins is a helper function that lists all DigitalTwin nodes.
+func (c *Client) ListDigitalTwins(
+	ctx context.Context,
+	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
+	return c.ListNodesByType(ctx, DigitalTwin, opts...)
+}
+
+// ListResources is a helper function that lists all Resource nodes.
+// that have the specified property.
+func (c *Client) ListResources(
+	ctx context.Context,
+	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
+	return c.ListNodesByType(ctx, Resource, opts...)
+}
+
+// ListNodesByType is a helper function that lists all nodes by type, regardless of whether they are DigitalTwins
+// or Resources. The nodeType argument should be in PascalCase.
+func (c *Client) ListNodesByType(
+	ctx context.Context,
+	nodeType string,
+	opts ...grpc.CallOption,
+) ([]*knowledgepb.Node, error) {
+	ctx = insertMetadata(ctx, c.xMetadata)
+	path := fmt.Sprintf("(:%s)", nodeType)
+	resp, err := c.client.IdentityKnowledge(ctx, &knowledgepb.IdentityKnowledgeRequest{
+		Path:      path,
+		Operation: knowledgepb.Operation_OPERATION_READ,
+	}, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return parseMultipleNodesFromPaths(resp.GetPaths())
+}
+
+func (c *Client) listNodesByProperty(
```

Comment:
Suggestion: in authorization we have `IsAuthorizedWithRawRequest` which basically is this method. But it's exposed in the SDK meaning the developer could potentially construct the payload himself without using wrapper methods.
Just a though to give more flexibility to the dev.

Reply:
Your suggestion is to expose this method, if I understand correctly?

Reply:
Yes, just a suggestion. It's nice to have helper methods, but also allow the developer to write the entire payload.

Reply:
`IdentityKnowledge.Read()` let's them do that. But no objection to also exposing the `listNodesByProperty` as well, but that would also be a helper.